workshop/session_3/11_biopython_proteins.ipynb

```diff
@@ -0,0 +1,957 @@
+    "ALNLHTANWFLYLSTLRWFQMRIEMIFVIFFIAVTFISILTTGEGEGRVGIILTLAMNIM\n",
+    "STLQWAVNSSIDVDSLMRSVSRVFKFIDMPTEGKPTKSTKPYKNGQLSKVMIIENSHVKK\n",
+    "DDIWPSGGQMTVKDLTAKYTEGGNAILENISFSISPGQRVGLLGRTGSGKSTLLSAFLRL\n",
+    "LNTEGEIQIDGVSWDSITLQQWRKAFGVIPQKVFIFSGTFRKNLDPYEQWSDQEIWKVAD\n",
+    "EVGLRSVIEQFPGKLDFVLVDGGCVLSHGHKQLMCLARSVLSKAKILLLDEPSAHLDPVT\n",
+    "YQIIRRTLKQAFADCTVILCEHRIEAMLECQQFLVIEENKVRQYDSIQKLLNERSLFRQA\n",
+    "ISPSDRVKLFPHRNSSKCKSKPQIAALKEETEEEVQDTRL\"\"\".replace(\"\\n\", \"\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "We will use a sliding window approach.\n",
+    "\n",
+    ">For a pre-defined window size of `n`, at any given point in the sequence, we will average its current and `(n-1)/2` preceding and proceeding values.\n",
+    "\n",
+    "You can think of it as a fixed-size rectangle moving across the sequence, averaging out the results to calculate the mean value of a wider section."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 167,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Calculated hydrophobicity for 1480 positions\n",
+      "Average hydrophobicity: 0.0318\n"
+     ]
+    }
+   ],
+   "source": [
+    "window_size = 11\n",
+    "\n",
+    "hydrophobicity = []  # The hydrophobicity value at a given point in the sequence.\n",
+    "for i, aa in enumerate(sequence):  # This will return tuple of a position in the sequence and the aminoacid.\n",
+    "    window_start = int(i - (window_size-1)/2)\n",
+    "    window_end = int(i + (window_size-1)/2)+1\n",
+    "\n",
+    "    if window_start < 0 or window_end > len(sequence):\n",
+    "        window_hydrophobicity = None  # At the very beginning and at the very end the window will be outside of the sequence.\n",
+    "\n",
+    "    else:\n",
+    "        aas_in_window = sequence[window_start:window_end]  # A list of all the aminoacids in the window.\n",
+    "        window_hydrophobicity = sum([Kyte_and_Doolittle_scale[aa] for aa in aas_in_window]) / window_size\n",
+    "    hydrophobicity.append(window_hydrophobicity)\n",
+    "\n",
+    "print(\"Calculated hydrophobicity for {} positions\".format(len(hydrophobicity)))\n",
+    "\n",
+    "# Note that this is slightly different than GRAVY from the reference paper.\n",
+    "print(\"Average hydrophobicity:\", \"{:.4f}\".format(sum(h if h else 0 for h in hydrophobicity) / len(hydrophobicity))) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "Now let's plot the hydrophobicity along the sequence to detect hydrophobic and hydrophilic regions. \n",
+    "\n",
+    "We will initialise `matplotlib` visualisation so that figures can be displayed in notebook cells.\n",
+    "\n",
+    "Then we will ask it to plot the `hydrophobicity` list from the previous calculation."
```

Comment:
https://en.wikipedia.org/wiki/Hydrophilicity_plot, They may find this link useful

Reply:
Added the link.